src/Microsoft.OData.Core/ODataParameterReaderCore.cs

```diff
@@ -192,7 +194,7 @@ public override sealed bool Read()
         public override sealed Task<bool> ReadAsync()
         {
             this.VerifyCanRead(false);
-            return this.ReadAsynchronously().FollowOnFaultWith(t => this.EnterScope(ODataParameterReaderState.Exception, null, null));
+            return this.InterceptExceptionAsync((thisParam) => thisParam.ReadAsynchronously());

```

Comment:
It's functionally equivalent for the happy case. However, when an exception is thrown `FollowOnFaultWith` is known to throw an `AggregateException` instead of the expected `ODataException` a somewhat flawed implementation (IMO) in `TaskUtils`. You'll recall we have had a few issues raised related to that... The approach here fixes those kind of bugs
```csharp
private static Task<TResult> FollowOnFaultWithImplementation<TResult>(Task antecedentTask, Func<Task, TResult> getTaskResult, Action<Task> operation)
{
    TaskCompletionSource<TResult> taskCompletionSource = new TaskCompletionSource<TResult>();
    antecedentTask.ContinueWith(
        t =>
        {
            switch (t.Status)
            {
                case TaskStatus.RanToCompletion:
                    taskCompletionSource.TrySetResult(getTaskResult(t));
                    break;
                case TaskStatus.Faulted:
                    try
                    {
                        operation(t);
                        taskCompletionSource.TrySetException(t.Exception);
                    }
                    catch (Exception exception)
                    {
                        if (!ExceptionUtils.IsCatchableExceptionType(exception))
                        {
                            throw;
                        }

                        AggregateException aggregateException = new AggregateException(t.Exception, exception);
                        taskCompletionSource.TrySetException(aggregateException);
                    }

                    break;
                case TaskStatus.Canceled:
                    taskCompletionSource.TrySetCanceled();
                    break;
            }
        },
        TaskContinuationOptions.ExecuteSynchronously)

        // This is only for the body of delegate in the .ContinueWith, the antecedent task failures are handles by the body itself.
        .IgnoreExceptions();
    return taskCompletionSource.Task;
}
```